ethereumj-core/src/main/java/org/ethereum/solidity/SolidityType.java

```diff
@@ -393,23 +388,73 @@ public String getCanonicalName() {
             } else {
                 throw new RuntimeException("Invalid value for type '" + this + "': " + value + " (" + value.getClass() + ")");
             }
-            return encodeInt(bigInt);
+            return bigInt;
+        }
+    }
+
+    public static class IntType extends NumericType {
+        public IntType(String name) {
+            super(name);
         }
 
+        @Override
+        public String getCanonicalName() {
+            if (getName().equals("int")) return "int256";
+            return super.getCanonicalName();
+        }
+        
+        public static BigInteger decodeInt(byte[] encoded, int offset) {
+            return new BigInteger(Arrays.copyOfRange(encoded, offset, offset + 32));
+        }
+        public static byte[] encodeInt(int i) {
+            return encodeInt(new BigInteger("" + i));
+        }
+        public static byte[] encodeInt(BigInteger bigInt) {
+            return ByteUtil.bigIntegerToBytesSigned(bigInt, 32);
+        }
         @Override
         public Object decode(byte[] encoded, int offset) {
             return decodeInt(encoded, offset);
         }
-
+        @Override
+        public byte[] encode(Object value) {
+            BigInteger bigInt = encodeInternal(value);
+            return encodeInt(bigInt);
+        }
+    }
+    
+    public static class UnsignedIntType extends NumericType {
+        public UnsignedIntType(String name) {
+            super(name);
+        }
+        
+        @Override
+        public String getCanonicalName() {
+            if (getName().equals("uint")) return "uint256";
+            return super.getCanonicalName();
+        }
         public static BigInteger decodeInt(byte[] encoded, int offset) {
-            return new BigInteger(Arrays.copyOfRange(encoded, offset, offset + 32));
+            return new BigInteger(1, Arrays.copyOfRange(encoded, offset, offset + 32));
         }
         public static byte[] encodeInt(int i) {
             return encodeInt(new BigInteger("" + i));
         }
         public static byte[] encodeInt(BigInteger bigInt) {
             return ByteUtil.bigIntegerToBytesSigned(bigInt, 32);
         }
+
+        @Override
+        public byte[] encode(Object value)
+        {
+            BigInteger bigInt = encodeInternal(value);
+            return encodeInt(bigInt);
```

Comment:
I'm not sure but maybe we should throw an exception here if `bigInt` < 0. Here and in encodeInt methods.
@mkalinin what do you think?

Reply:
That makes sense to me 👍